### DISCUSSION TITLE
Why do I get an Uncaught Exception when using a stream provider?

### DISCUSSION THREAD

**Discussion by eladm-ultrawis** [rrousselGit/riverpod#3717] [Q&A]
I have a provider that is listening to a broadcast stream (originating from a grpc stream)

```
final serverNotificationsStreamProvider =
    StreamProvider<ultraproto.ChangeNotificationType>((ref) async* {
  final stream = await ref.watch(getServerCommandsStreamProvider.future);

  await for (final val in stream) {
    if (val.hasChangeNotification()) {
      final type = val.changeNotification.type;
      yield val.changeNotification.type;
      switch (type) {
        case ultraproto.ChangeNotificationType.CHANGE_NOTIFICATION_TYPE_SITE:
          yield val.changeNotification.type;
        default:
          logger.e(LoggingCategory.webGrpcClient,
              'Unknown change notification type');
          break;
      }
    }
  }
});
```

when certain data changes, the stream is cancelled. hence the `await for` loop throws an exception.
This makes sense and is what I expect. However, when debugging this with vscode, it seems that this exception is considered "Uncaught Exception". I would expect the StreamProvider mechanism to catch this exception and report this via the AsyncValue mechanism. 

my questions:

1. is there a way to properly handle the exceptions emanating from the code inside the StreamProvider?
2. is this the expected behavior?

**Comment by snapsl** [comment]
Hi,

>  I would expect the StreamProvider mechanism to catch this exception and report this via the AsyncValue mechanism.

Yes, it should! This could be a bug. Perhaps you can fill an issue with a reproducible example.
